### PR TITLE
[mlir][tensor] `tensor.generate`: do not verify dynamic sizes

### PR DESCRIPTION
**Pull request by matthias-springer**
Op verifiers should verify only local properties of an op. The dynamic sizes of a `tensor.generate` op should not be verified. Dynamic sizes that have a negative constant value should not prevent the `tensor.generate` op from verifying.

Also share some code between the `tensor.empty` and `tensor.generate` "dynamic dim -> static dim" canonicalization patterns.

Remove the `invalid-canonicalize.mlir` file and move the test case to `canonicalize.mlir`. Canonicalization no longer produces IR that does not verify (and leaves the op as is).

### CODE REVIEW

File: mlir/test/Dialect/Tensor/canonicalize.mlir [hunk 1]
```diff
@@ -2018,3 +2018,23 @@ func.func @invalid_slice_ops(%t: tensor<?xf32>, %t2: tensor<?xf32>) -> tensor<?x
   %1 = tensor.insert_slice %0 into %t2[2][%c][1] : tensor<?xf32> into tensor<?xf32>
   return %1 : tensor<?xf32>
 }
+
+// -----
+
+// Negative dynamic dimension sizes should not be turned into static sizes.
+// They should also not prevent the op from verifying.
```
Comment (rikhuijzer) [inline mlir/test/Dialect/Tensor/canonicalize.mlir:2025]: Isn't the convention to describe test purpose in test func name and not do comments? When confused about a test, then git blame usually has all the information anyway.

**Review by rikhuijzer — APPROVED** [review]
Apart from the nit comment it LGTM.

File: mlir/test/Dialect/Tensor/canonicalize.mlir (hunk 1 above)
Comment (matthias-springer) [reply mlir/test/Dialect/Tensor/canonicalize.mlir:2025]: Personally, I'd be happy to see more comments in unit test cases. We have quite a few of those in the bufferization dialect unit tests.

Comment (rikhuijzer) [reply mlir/test/Dialect/Tensor/canonicalize.mlir]: Fair enough.

**Pull request merged**